python/lsst/pipe/base/graphBuilder.py

```diff
@@ -191,22 +126,143 @@ def dimensions(self) -> DimensionGraph:
+        """Unpack nested single-element `DatasetRef` dicts into a new
+        mapping with `DatasetType` keys and `DatasetRef` values.
 
-    def unpackRefs(self) -> NamedKeyDict:
-        """Unpack nested single-element `DatasetRef` lists into a new
-        dictionary.
+        This method assumes that each nest contains exactly one item, as is the
+        case for all "init" datasets.
 
-        This method assumes that each `_DatasetScaffolding.refs` list contains
-        exactly one `DatasetRef`, as is the case for all "init" datasets.
+        Returns
+        -------
+        dictionary : `NamedKeyDict`
+            Dictionary mapping `DatasetType` to `DatasetRef`, with both
+            `DatasetType` instances and string names usable as keys.
+        """
+        def getOne(refs: Dict[DataCoordinate, DatasetRef]) -> DatasetRef:
+            ref, = refs.values()
+            return ref
+        return NamedKeyDict({datasetType: getOne(refs) for datasetType, refs in self.items()})
+
+    def unpackMultiRefs(self) -> NamedKeyDict[DatasetType, DatasetRef]:
+        """Unpack nested multi-element `DatasetRef` dicts into a new
+        mapping with `DatasetType` keys and `set` of `DatasetRef` values.
 
         Returns
         -------
         dictionary : `NamedKeyDict`
             Dictionary mapping `DatasetType` to `DatasetRef`, with both
             `DatasetType` instances and string names usable as keys.
         """
-        return NamedKeyDict((datasetType, scaffolding.refs[0]) for datasetType, scaffolding in self.items())
+        return NamedKeyDict({datasetType: list(refs.values()) for datasetType, refs in self.items()})
+
+    def extract(self, datasetType: DatasetType, dataIds: Iterable[DataCoordinate]
+                ) -> Iterator[DatasetRef]:
+        """Iterate over the contained `DatasetRef` instances that match the
+        given `DatasetType` and data IDs.
+
+        Parameters
+        ----------
+        datasetType : `DatasetType`
+            Dataset type to match.
+        dataIds : `Iterable` [ `DataCoordinate` ]
+            Data IDs to match.
+
+        Yields
+        ------
+        ref : `DatasetRef`
+            DatasetRef instances for which ``ref.datasetType == datasetType``
+            and ``ref.dataId`` is in ``dataIds``.
+        """
+        refs = self[datasetType]
+        return (refs[dataId] for dataId in dataIds)
+
+
+@dataclass
+class _QuantumScaffolding:
+    """Helper class aggregating information about a `Quantum`, used when
+    constructing a `QuantumGraph`.
+
+    See `_PipelineScaffolding` for a top-down description of the full
+    scaffolding data structure.
+
+    Parameters
+    ----------
+    task : _TaskScaffolding
+        Back-reference to the helper object for the `PipelineTask` this quantum
+        represents an execution of.
+    dataId : `DataCoordinate`
+        Data ID for this quantum.
+    """
+    def __init__(self, task: _TaskScaffolding, dataId: DataCoordinate):
```

Comment:
I think taking that approach (along with the changes in your next few comments) also requires adding `field(init=False)` or defaults (and hence dropping `__slots__`) for `inputs`, `outputs`, and `prerequisites`, and that `task` _shouldn't_ actually be `InitVar` in that scenario since it's also an attribute.  Overall it seems simpler to just define my own `__init__` than go to so much work to make the `dataclass-injected` one usable for me.

Maybe the better question is whether I should drop `@dataclass` - I'm already overriding `__repr__` and `__init__`, so all I'm getting from that is equality comparison, and I don't actually want that, I suppose.  Unless I hear otherwise I'll just drop the decorator.